### PR TITLE
Validate Issuer Using New Validation Model in Saml2SecurityTokenHandler

### LINKED ISSUE #2880: Remove Pragma Disable Warning For New ValidateTokenAsync Method on Saml2TokenHandler
Due to the TDD approach we are taking on constructing the new ValidateTokenAsync method on Saml2TokenHandler that implement our new token validation model, we introduced `pragma warnings disable` around methods that passed CancellationToken. We should be able to remove them once we get into async validators such as the Issuer validator.

### LINKED ISSUE #2883: Implement and Test Issuer Validation in Saml2 Token Handler Using New Token Validation Model
Add Issuer validation to new ValidateTokenAsync() flow in both SamlSecurityTokenHandler and Saml2SecurityTokenHandler.

### PR DESCRIPTION
**Pull request by FuPingFranco**
# Validate Issuer Using New Validation Model in Saml2SecurityTokenHandler

## Description

Implements issuer validation as part of the new ValidateTokenAsync method in Saml2SecurityTokenHandler

Fixes #2883 and #2880

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2NameIdentifier.cs [hunk 1]
```diff
@@ -104,7 +104,7 @@ public string Value
             get { return _value; }
             set
             {
-                if (string.IsNullOrEmpty(value))
+                if (string.IsNullOrEmpty(value)) //NOTE: We can remove this check and let our issuer validator handle this.
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2NameIdentifier.cs:107]: With regards to the note, `Saml2NameIdentifier` is used by the issuer, but that is not the only instance.
Unfortunately, it looks like we will need to look more into this as part of the **reading** of the SAML2 token.

I'll add a separate issue to refactor the token reading like we did for JWT.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs [hunk 2]
```diff
@@ -40,13 +37,31 @@ internal async Task<ValidationResult<ValidatedToken>> ValidateTokenAsync(
...
             }
 
+            Task<ValidationResult<ValidatedIssuer>> validatedIssuerResult = validationParameters.IssuerValidatorAsync(
+                samlToken.Issuer,
+                samlToken,
+                validationParameters,
+                callContext,
+                cancellationToken);
+
+            if (!(await validatedIssuerResult.ConfigureAwait(false)).IsValid)
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs]: Awaiting the result does not seem right. I'll look into this.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs [hunk 3]
```diff
@@ -40,13 +37,31 @@ internal async Task<ValidationResult<ValidatedToken>> ValidateTokenAsync(
...
+                samlToken,
+                validationParameters,
+                callContext);
 
             if (!conditionsResult.IsValid)
             {
                 return conditionsResult.UnwrapError().AddStackFrame(new StackFrame(true));
             }
 
+            Task<ValidationResult<ValidatedIssuer>> validatedIssuerResult = validationParameters.IssuerValidatorAsync(
```
Comment (iNinja) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs]: This should be `ValidationResult<ValidatedIssuer> validatedIssuerResult = await validationParameters.IssuerValidatorAsync(...`

**Comment by jennyf19** [comment]
@FuPingFranco will you be addressing the comments

**Comment by iNinja** [comment]
@FuPingFranco please advise if you intend to update this, or I can make the changes.
I just want to avoid duplicating efforts.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2NameIdentifier.cs (hunk 1 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2NameIdentifier.cs:107]: This is related to this issue https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2905 as well if you'd like to include it into a new issue.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs (hunk 2 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs]: Updated. Thanks @iNinja

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs (hunk 3 above)
Comment (FuPingFranco) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs]: Updated. Thanks @iNinja

**Comment by FuPingFranco** [comment]
@jennyf19 & @iNinja , I addressed the comments.

**Review by iNinja — APPROVED** [review]

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlTokenUtilities.cs [hunk 4]
```diff
@@ -154,5 +154,35 @@ internal static async Task<TokenValidationParameters> PopulateValidationParamete
...
+
+            validationParametersCloned.ValidIssuers.Add(currentConfiguration.Issuer);
+
+            foreach (SecurityKey key in currentConfiguration.SigningKeys)
+            {
+                validationParametersCloned.IssuerSigningKeys.Add(key);
+            }
+
+            return validationParametersCloned;
+
```
Comment (kllysng) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlTokenUtilities.cs]: Super nit 🦹 
```suggestion
```

**Review by kllysng — APPROVED** [review]

**Review by westin-m — APPROVED** [review]

**Pull request merged**